### DISCUSSION TITLE
New reducer initialises 4 times

### DISCUSSION THREAD

**Discussion by katovBash** [pointfreeco/swift-composable-architecture#1584] [General]
### Description

When I'm opening my module that contains different reducers, I have noticed that my new reducers trigger 4 times, but my old one only one time, how I can fix this behaviour? 

This is how I'm using my new reducer with old one. Is it bad case to use AnyReducer?

```
private let childrenReducer = MainReducerType.combine(

AnyReducer { state, action, _ in
    DynamicReducer().reduce(into: &state, action: action)
  }
  .optional()
  .pullback(
    state: \MainState.tvState,
    action: /MainActions.tvActions,
    environment: id
  ),
  
  ...
  )
```

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Initialise 1 time

### Actual behavior

Initialise 4 times

### Steps to reproduce

Open module where you have combine reducers, check how many times inits old reducer and new reducer. 

### The Composable Architecture version information

0.42.0

### Destination operating system

iOS 16

### Xcode version information

14.0

**Comment by mbrandonw** [comment]
Hi @katovBash, can you explain a bit more of what the problem is? Is the problem that the reducer is _created_ 4 times, or that its logic is _executing_ 4 times?

Also I am going to convert this to a discussion since I do not believe it is an issue with the library.

**Reply by katovBash** [reply]
Hi!

I have reducer that contains another reducers)

I have old reducer from 1.0 like:

```
ChatReducer
    .optional()
    .pullback(
      state: \MainState.chatState,
      action: /MainActions.chatActions,
      environment: \.chatEnvironment
    ),
```
    
It works correctly, initialise only one time. 
 
 
 Also, I have a new reducer like:
 ```
AnyReducer { state, action, _ in
    DynamicReducer().reduce(into: &state, action: action)
  }
  .optional()
  .pullback(
    state: \MainState.tvState,
    action: /MainActions.tvActions,
    environment: id
  )
```
I put pb in the scope of AnyReducer and it triggers 4 time, but it should execute only one time. On appear executes correctly.

**Reply by mbrandonw** [reply]
Sometimes breakpoints can be flakey in Xcode. If you put a print statement in `DynamicReducer`'s `reduce` method, can you confirm it prints 4 times?

**Reply by katovBash** [reply]
yes, I put print in scope of anyReducer and it prints 4 times :-(

**Reply by mbrandonw** [reply]
I think you need to share more code for anyone to help. There's nothing in the code to explain why it is being called 4 times, and I don't think this is a problem in the library itself.

**Reply by stephencelis** [reply]
You might also add a `._printChanges()` to your reducer to see how many actions are send to the store. With the `body` style of reducer it's expected that they will be initialized once per action and not once for the duration of the application, as was done before the protocol. This behavior is expected, though, and aligns with how SwiftUI views are initialized, as well.

**Reply by katovBash** [reply]
The quantity of calling actions are also correct. On appear in reducer actions also trigger only once as expected. 

It seems that it triggers 4 times only in scope of anyReducer. 

I wouldn't pay attention of it, but I have some dependency to my module that I don't need in every dynamicReducer and that can be change from reducer to reducer. That's why I initialise it in this way: 

```
AnyReducer { state, action, _ in
    DynamicReducer(
      dynamicUpdaterService: DynamicUpdaterService(updater: DynamicUserUpdater())
    )
    .reduce(into: &state, action: action)
  }
  .optional()
  .pullback(
    state: \MainState.shopState,
    action: /MainActions.shopActions,
    environment: id
  ),
```

As u can expect my dynamicUserUpdater has 4 instances of it :(

Then what's a better approach to make this di if in scope of AnyReducer it triggers 4 times? 

Is it a good approach to use it in @Dependency?
I think that its not the best way because I need this dependency only for a few modules)

**Reply by stephencelis** [reply]
@katovBash Yup! Global dependencies should be held in `DependencyValues` if you want to ensure they live for the duration of the app.

**Reply by katovBash** [reply]
@stephencelis 

Ok. Thanks. But when I want to use it only in particular module, it's a good approach to use it like:

AnyReducer { state, action, _ in
    DynamicReducer(
      dynamicUpdaterService: DynamicUpdaterService(updater: DynamicUserUpdater())
    )
    .reduce(into: &state, action: action)
  }
  
  Initialise it only in the particular module? Again, its connected with code upper.
